signetlabdei/release/0.4

neerajvarshney opened the pull request:
This PR considers following major items
1. Two QD models (NIST/TGay)
2. Several indoor and outdoor example folders
3. Updated documentation

neerajvarshney committed:
examples added

neerajvarshney committed:
merged

neerajvarshney committed:
Error fixed in tree travesal

For thired order reflections

neerajvarshney committed:
boxLectureRoom Node Positions

Node positions changed

neerajvarshney committed:
all examples file checked

neerajvarshney committed:
flowchart integration

neerajvarshney committed:
Update parameterCfg.m

neerajvarshney committed:
material lib files

neerajvarshney committed:
merged checked again

neerajvarshney committed:
test suite failed

neerajvarshney committed:
test suite failed

neerajvarshney committed:
review comments addressed

neerajvarshney committed:
Update parameterCfg.m

neerajvarshney committed:
city block example

neerajvarshney committed:
test suite failed

neerajvarshney committed:
test suite passed

neerajvarshney committed:
testSuite.m

neerajvarshney committed:
cityblock JSON output

neerajvarshney committed:
Code cleaning

neerajvarshney committed:
Document, Code cleaning, re-run scenarios

mattia-lecci commented on src/examples/boxLectureRoom/Input/Box.xml:
```diff
@@ -0,0 +1,310 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
keep the examples' names consistent: all other scenarios start with a capital letter

neerajvarshney replied:
This is suggested by our JAMES BOND "Tanguy" so please keep it like this. Scenario name  in each of the string is consistent e.g., ParkingLot, simplifiedParkingLot, LivingRoom, etc.

tangnist replied:
If I suggested that, then I'm afraid I'm going to have to disagree with myself. I probably wrote the scenario name this way as I didn't know all the others were capitalized at the beginning. So I'm for consistency so "BoxLectureRoom"

neerajvarshney replied:
done! I have changed folder name and xml file name as suggested.

mattia-lecci replied:
hi, the fix was probably not pushed? I still see "boxLectureRoom" and "simplifiedParkingLot", as well as "BoxLectureRoom/Input" [here](https://github.com/usnistgov/qd-realization/tree/feature/morescenarios/src/examples)

neerajvarshney replied:
![image](https://user-images.githubusercontent.com/58748792/106259937-dbefc780-61ed-11eb-9bdf-52e4c2aa9d13.png)

neerajvarshney replied:
i do not have these folders

neerajvarshney replied:
how to resolve this?

neerajvarshney replied:
resolved! check now

mattia-lecci commented on src/examples/simplifiedParkingLot/Input/simplifiedParkingLot.xml:
```diff
@@ -0,0 +1,1375 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
same as above

neerajvarshney replied:
This is suggested by our JAMES BOND "Tanguy" so please keep it like this. Scenario name  in each of the string is consistent e.g., ParkingLot, simplifiedParkingLot, LivingRoom, etc.

tangnist replied:
If I suggested that, then I'm afraid I'm going to have to disagree with myself. "SimplifiedParkingLot.xml"

neerajvarshney replied:
done

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -72,17 +68,39 @@
+% as -inf which means software does not discard any diffuse components
+para = fieldToNum(para, 'diffusePathGainThreshold', [], -inf);
+
+% Qausi deterministic model
```
*Quasi-deterministic

neerajvarshney replied:
done

mattia-lecci replied:
*Quasi

neerajvarshney replied:
done

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -72,17 +68,39 @@
+% nistMeasurements : model based on NIST measurements (Default), 
+% tgayMeasurements : model based on TGay channel document measurements. 
+if ~isfield(para, 'switchQDModel')
+    warning(strcat('Q-D model is not defined in paraCfgCurrent.txt. ',...
+        ' Thus, considering nistMeasurements (default) switchQDModel.'));
```
```suggestion
    warning(['Q-D model is not defined in paraCfgCurrent.txt. ',...
        'Thus, considering nistMeasurements (default) switchQDModel.']);
```

neerajvarshney replied:
done

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -72,17 +68,39 @@
 para = fieldToNum(para, 'totalNumberOfReflections', [], 2);
+if strcmp(para.switchQDModel,'tgayMeasurements') ...
+        && para.totalNumberOfReflections>2
+    warning(strcat('totalNumberOfReflections for switchQDModel = tgayMeasurements',...
+        'can not be considered higher than 2. Thus, setting Default value (2).'));
```
```suggestion
    warning(['totalNumberOfReflections for switchQDModel = tgayMeasurements ',...
        'can not be considered higher than 2. Thus, setting Default value (2).']);
```

neerajvarshney replied:
done

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -72,17 +68,39 @@
 % Order of reflection.
 % 1 = multipath until first order, 2 = multipath until second order (Default)
 para = fieldToNum(para, 'totalNumberOfReflections', [], 2);
+if strcmp(para.switchQDModel,'tgayMeasurements') ...
```
any reason why are you constraining the reflection order? Why can't the user try higher reflection orders?

neerajvarshney replied:
Yes we are calculating angle of incident till second order of reflection in TGAY model. This has been clearly described in the document

mattia-lecci replied:
ok, thanks for the clarification.
In this case, i'm not sure that a warning is enough.
I had discussion from last year-ish with @tangnist (not sure why github won't let me mention him directly..) were we had different opinions on whether we should punish a user for a poor decision (e.g., with an error) or correct the poor behavior assuming that you know better than him and possibly notifying him (e.g., with a warning).
There a pros and cons for both sides, if a warning is ok for you guys then i'm fine with it, I just wanted to point this out

neerajvarshney replied:
He did not pointed out anything in documentation so I think he should be fine with the warning.

tangnist replied:
Mattia, I'm still in general more in favor of punishment, but at one point, we need to deliver so I'm fine with it! Moreover, when the behavior is described in the documentation,  at least our user still gets a chance to understand what happened.

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -108,12 +126,28 @@
     end
+else
+    currentPath =pwd;
+    materialLibraryAbsPath = fullfile(currentPath(1:find((pwd==filesep)==1, ...
+        1, 'last')), 'src',para.materialLibraryPath); % For test suite
```
check https://www.mathworks.com/help/matlab/ref/fileparts.html

neerajvarshney replied:
done

mattia-lecci commented on src/parameterCfg.m:
```diff
@@ -108,12 +126,28 @@
+    if ~isfile(materialLibraryAbsPath)...
+        || ~isMaterialLibraryFileFormat(para.switchQDModel,...
+        materialLibraryAbsPath,para.materialLibraryPath)
+        warning(strcat('Using Empty material library. Check following issues: ',...
+            ' 1. Material file not exist. ',...
+            ' 2. Material file format not correct.'));
```
check spaces

neerajvarshney replied:
I did ctr + I, if you are referring something else, let me know

neerajvarshney replied:
warning(['Using Empty material library. Check following issues: ',...
            '1. Material file not exist. ',...
            '2. Material file format not correct.']);

neerajvarshney replied:
this is what you want?
I checked both has same spacing

mattia-lecci replied:
I was referring to printed spaces: you have white spaces both at the beginning and at the end of each line, which will be translated into a double space one it gets printed.
To keep the consistency, please keep the white space at the end of the line and remove it from the beginning

neerajvarshney replied:
corrected

mattia-lecci commented on src/raytracer/Raytracer.m:
```diff
@@ -56,10 +57,10 @@
         paraCfgInput.carrierFrequency ~= 60e9
     warning(['Please, note that diffuse scattering model is only ',...
-        'valid for fc=60 GHz'])
+        'valid for fc=60 or 28 GHz'])
```
modify if statement accordingly.
more in general this could even be removed: different material libraries consider different frequencies and at the moment this cannot be checked

neerajvarshney replied:
True, I removed this if statement completely

mattia-lecci commented on src/raytracer/getTgayReflectionLoss.m:
```diff
@@ -0,0 +1,107 @@
+function reflectionLossdB = getTgayReflectionLoss(MaterialLibrary,...
+                materialIndex, multipath)
+% getTgayReflectionLoss returns reflection loss in dB. 
```
missing documentation.
also a reference would be nice

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/getTgayReflectionLoss.m:
```diff
@@ -0,0 +1,107 @@
+end     
+end
+
+function incidentAngle = angleOfIncident(multipath) 
```
*angleOfIncidence
missing doc

mattia-lecci replied:
not fixed?

neerajvarshney replied:
you mean output variable?

neerajvarshney replied:
okay! you are about function name! wait let me correct

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/getTgayReflectionLoss.m:
```diff
@@ -0,0 +1,107 @@
+reflectionLossdB = -20*log10(abs(reflectionLoss)); 
+end
+
+function reflectionCoefficient = getReflectance(MaterialLibrary,...
```
missing doc

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/qdGenerator.m:
```diff
@@ -1,7 +1,10 @@
 function [output,  outputPre, outputPost] =...
```
update documentation

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/verifyPath.m:
```diff
@@ -5,58 +5,64 @@
+% instead of custom ones, supporting mobility check, improved if/else chain
 
 
 switch3 = 1;
```
rename

neerajvarshney replied:
done

mattia-lecci replied:
i'd suggest pathExists instead of isPathExist

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/verifydistance.m:
```diff
@@ -1,42 +1,42 @@
 function switch1 = verifydistance(r, Tx, CADop, i)
```
missing doc
rename switch1

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/verifypathTreeTraversal.m:
```diff
@@ -30,40 +30,41 @@
 %switch3 - boolean which holds the information of possibility of path
```
rename

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/xmlreader.m:
```diff
@@ -1,49 +1,52 @@
-%The output is extracted triangles (CADop),
-%and a boolean to know whether the material information is
-%present in the CAD file (switch1)
+%XML redaer extracts the information of CAD file (AMF). 
+% The input of the function is filename, the material database with all the 
+% material parameters (Material_library), reference point (referencePoint)
+% and distance limitation(r)
+% The output is extracted triangles (CADop),
+% and a boolean to know whether the material information is
+% present in the CAD file (switch1)
```
improve documentation with the usual format

neerajvarshney replied:
done

mattia-lecci commented on src/raytracer/xmlreader.m:
```diff
@@ -186,7 +189,7 @@
             
             % We are using distance limitation at this step
-            if r==0
+            if r==inf
```
use isinf(r), it should be safer

neerajvarshney replied:
done

mattia-lecci commented on src/utils/getCadOutput.m:
```diff
@@ -54,7 +54,8 @@
     else
-        warning('Cache file exists but might be outdated. Ignoring it.')
+        warning(strcat('Cache file exists but might be outdated. ',...
+            ' Ignoring it and generating new one.'));
```
check spaces

neerajvarshney replied:
which spaces. I did ctr+I, let me know if you are talking about something else

neerajvarshney replied:
warning(['Cache file exists but might be outdated. ',...
            'Ignoring it and generating new one.']);

neerajvarshney replied:
I have changed it as above

mattia-lecci reviewed (commented):
If I understand correctly, the examples were updated as follows:
- added CityBlock
- added HotelLobby
- removed Indoor1
- removed Sphere
- added LectureRoom
- added LivingRoom
- added OpenAreaHotspot
- added ParkingLot
- added boxLectureRoom
- added simplifiedParkingLot
- removed Outdoor1
(plus some fixes to already existing scenarios)

In general the review is just a minor, good job :)

neerajvarshney commented:
> If I understand correctly, the examples were updated as follows:
> 
> * added CityBlock
> * added HotelLobby
> * removed Indoor1
> * removed Sphere
> * added LectureRoom
> * added LivingRoom
> * added OpenAreaHotspot
> * added ParkingLot
> * added boxLectureRoom
> * added simplifiedParkingLot
> * removed Outdoor1
>   (plus some fixes to already existing scenarios)
> 
> In general the review is just a minor, good job :)

Yes! Thanks

neerajvarshney committed:
Padova Comments

neerajvarshney committed:
Box.xml renamed to BoxLectureRoom.xml

1. Box.xml renamed to BoxLectureRoom.xml for consistency
2. Removed Rotation [0 0 0] files from each examples (useless)
3. updated document for boxLectureRoom.xml

mattia-lecci commented on test/exampleOutputsTest.m:
```diff
@@ -91,15 +96,15 @@ function teardown(testCase)
 % boxLectureRoom
-function boxLectureRoomTest(testCase)
-exampleName = 'boxLectureRoom';
+function BoxLectureRoomTest(testCase)
```
start with lower case for consistency

neerajvarshney replied:
done

neerajvarshney replied:
done

mattia-lecci reviewed (commented):
almost there!
please check the answers to the old comments as well

neerajvarshney commented:
I have checked and addressed!

neerajvarshney committed:
Code cleaning

neerajvarshney committed:
removed BoxLectureRoom and SimplifiedParkingLot

neerajvarshney committed:
added BoxLectureRoom and SimplifiedParkingLot
